docker-compose-rule-junit4/src/main/java/com/palantir/docker/compose/DockerComposeRule.java

```diff
@@ -133,34 +152,78 @@ protected LogCollector logCollector() {
         return new DoNothingLogCollector();
     }
 
-    @Override
     public void before() throws IOException, InterruptedException {
         log.debug("Starting docker-compose cluster");
+
+        pullBuildAndUp();
+
+        logCollector().startCollecting(dockerCompose());
+
+        emitEventsFor().waitingForServices(this::waitForServices);
+    }
+
+    private void pullBuildAndUp() throws IOException, InterruptedException {
         if (pullOnStartup()) {
-            dockerCompose().pull();
+            emitEventsFor().pull(dockerCompose()::pull);
         }
 
-        dockerCompose().build();
+        emitEventsFor().build(dockerCompose()::build);
 
         DockerCompose upDockerCompose = dockerCompose();
         if (removeConflictingContainersOnStartup()) {
             upDockerCompose = new ConflictingContainerRemovingDockerCompose(upDockerCompose, docker());
         }
-        upDockerCompose.up();
 
-        logCollector().startCollecting(dockerCompose());
+        emitEventsFor().up(upDockerCompose::up);
+    }
+
+    private void waitForServices() throws InterruptedException {
         log.debug("Waiting for services");
-        new ClusterWait(ClusterHealthCheck.nativeHealthChecks(), nativeServiceHealthCheckTimeout())
-                .waitUntilReady(containers());
-        clusterWaits().forEach(clusterWait -> clusterWait.waitUntilReady(containers()));
+        Consumer<Cluster> nativeHealthCheckClusterWait =
+                emitEventsFor().nativeClusterWait(
+                        new ClusterWait(ClusterHealthCheck.nativeHealthChecks(), nativeServiceHealthCheckTimeout()));
+
+        List<Consumer<Cluster>> allClusterWaits = Stream.concat(
+                Stream.of(nativeHealthCheckClusterWait),
+                clusterWaits().stream().map(emitEventsFor()::userClusterWait))
+                .collect(Collectors.toList());
+
+        waitForAllClusterWaits(allClusterWaits);
+
         log.debug("docker-compose cluster started");
     }
 
-    @Override
+    private void waitForAllClusterWaits(List<Consumer<Cluster>> allClusterWaits) throws InterruptedException {
+        ListeningExecutorService executorService = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(
+                allClusterWaits.size(),
```

Comment:
do we need a thread per consumer?

Reply:
Yep, we need one thread per cluster wait. Since we only know when a service becomes healthy by checking, if we check them serially a service that becomes healthy quickly could seem like it takes forever as it's checked after a slow starting service has become healthy.

Reply:
Welp, I tested this on an internal product JT that has 9 services in it's docker compose and it actually fails to startup due to increased CPU usage using this version of DCR, probably from all polling. I think we need to do something like use a higher pollInterval (maybe it becomes every 1 sec after 2 secs of trying?)  or a `ScheduledExecutorService` instead to get it back below the timeout.

Reply:
Now poll every 1 second to reduce CPU load.